### PR TITLE
Rw 174

### PR DESCRIPTION
**Pull request by attiks**

### CODE REVIEW

**Review by orakili — COMMENTED** [review]
@attiks I wonder how much of the ModerationServiceBase can be reused for the My Posts page.

The service was designed with the assumption of handling a single bundle and with a particular format for the displayed table and filters that the Editors are trained to use.

The My Posts page is however mostly intended for the job/training posters (different than the editors) and the table and filters are mostly different than the moderation pages.

It may be easier to directly translate https://github.com/UN-OCHA/rwint-site/blob/master/site/profiles/reliefweb/modules/features/reliefweb_opportunities/includes/user-posts.inc to D9 code than re-using the moderation service.

File: html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    return $rows;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function initFilterDefinitions(array $filters = []) {
+    $definitions = parent::initFilterDefinitions([
```
Comment (orakili) [inline html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php:130]: Need a filter on the entity ID.

**Review by orakili — COMMENTED** [review]
@attiks That looks promising, the changes to the `ModerationServicBase` are much lighter now and sensible. 

I think having the `title`, `id` and `source` in the "omnibox" is a good compromise.

File: html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      'content_type',
+      'title',
+      'status',
+      'job_closing_date',
```
[the remark this answers is not in the document]
Comment (attiks) [reply html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php]: removed

File: html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+      'date' => [
+        'label' => $this->t('Posted'),
+        'type' => 'property',
+        'specifier' => 'created',
+        'sortable' => TRUE,
+      ],
+      'deadline' => [
+        'label' => $this->t('Deadline'),
+        'type' => 'property',
+        'specifier' => 'deadline',
```
[the remark this answers is not in the document]
Comment (attiks) [reply html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php:98]: Deadline filtering added, but sorting is not working

**Review by orakili — COMMENTED** [review]
@attiks Looking good.

For the sorting of the headline, I wonder if we cannot do something like that (untested):

```php
  /**
   * {@inheritdoc}
   */
  protected function wrapQuery(Select $query, $limit = 30) {
    // Get the order information.
    $info = $this->getOrderInformation();
    $sort_direction = $info['sort'] ?? 'desc';

    // Special handling of the headline sorting.
    $headline_alias = '';
    if (isset($info['order']) && $info['order'] === 'headline') {
      // Check if the join for the headline field was already performed.
      foreach ($query->getExpressions() as $expression) {
        if (strpos($expression['alias'], 'headline') === 0) {
          $headline_alias = $expression['alias'];
          break;
        }
      }
      // If not, join the tables.
      if (empty($headline_alias)) {
        // Entity information.
        $entity_type_id = $this->getEntityTypeId();
        $entity_base_table = $this->getEntityTypeDataTable($entity_type_id);
        $entity_id_field = $this->getEntityTypeIdField($entity_type_id);

        // Filter definition.
        $definition = $this->getFilterDefinitions()['headline'];

        // Join the deadline tables.
        $headline_alias = $this->joinDeadline($query, $definition, $entity_type_id, $entity_base_table, $entity_id_field);
      }
    }

    // Let the parent wrap the query.
    $wrapper = parent::wrapQuery($query, $limit);

    // Add the sort property to the wrapper query.
    if (!empty($headline_alias)) {
      $wrapper->orderBy("subquery.{$headline_alias}", $sort_direction);
    }

    return $wrapper;
  }
  ```
  
For this to work, on line 97 (`::getHeaders()`), add `'type' => 'custom',` for example to ensure the headline field is not processed like the "normal" headers in `::getQueryOrderInformation()`.

File: html/modules/custom/reliefweb_user_posts/src/Form/UserPostsPageFilterForm.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+/**
+ * Content moderation page filter form handler.
+ */
+class UserPostsPageFilterForm extends ModerationPageFilterForm {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFormId() {
+    return 'reliefweb_moderation_page_filter_form';
```
Comment (orakili) [inline html/modules/custom/reliefweb_user_posts/src/Form/UserPostsPageFilterForm.php:20]:
```suggestion
    return 'reliefweb_user_posts_page_filter_form';
```

@attiks I think this should be changed.

That would notably allow us to add a template override: `form--reliefweb-user-posts-page-filter-form.html.twig` for Lisa to add her magic.

Comment (attiks) [reply html/modules/custom/reliefweb_user_posts/src/Form/UserPostsPageFilterForm.php:20]: If I change it, the javascript stops working

Comment (orakili) [reply html/modules/custom/reliefweb_user_posts/src/Form/UserPostsPageFilterForm.php:20]: @attiks I think adding `$form['#attributes']['id'] = 'reliefweb-moderation-page-filter-form';` at end of the `::buildForm()` would make the javascript work.

File: html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+      ]);
+      $cells['poster'] = $this->currentUser->id() == $entity->getOwner()->id() ? $this->t('me') : $this->t('other');
+      $cells['source'] = '';
+      $cells['title'] = $entity->toLink()->toString();
+      $cells['date'] = $this->getEntityCreationDate($entity);
+
+      $cells['deadline'] = '';
+      if ($entity->field_registration_deadline) {
+        $cells['deadline'] = $this->formatDate($entity->field_registration_deadline->value);
+      }
```
Comment (orakili) [inline html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php:135]: Note: it's allowed for a training to have an empty `field_registration_deadline`: for **ongoing** training. In that case the deadline label should be set to "`$cells['deadline'] = $this->t('Ongoing')`". So probably need to check the entity bundle in addition to the field value.

File: html/modules/custom/reliefweb_user_posts/src/Form/UserPostsPageFilterForm.php (hunk 4 above)
Comment (attiks) [reply html/modules/custom/reliefweb_user_posts/src/Form/UserPostsPageFilterForm.php]: Fixed

File: html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php (hunk 5 above)
Comment (attiks) [reply html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php]: Fixed, but ugly, for some reason `$entity->field_registration_deadline->value` returns the string `Array`, so `IsEmpty()` does not work

Comment (orakili) [reply html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php]: @attiks Would you have the ID of a node for which this happen?

Comment (attiks) [reply html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php]: I just created an ongoing training and opened `/moderation/content/training`

File: html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+      'operator' => 'AND',
+    ];
+
+    $definitions['author2'] = [
+      'type' => 'property',
+      'field' => 'uid',
+      'label' => $this->t('Posted by'),
+      'shortcut' => 'a',
+      'form' => 'other',
+      'join_callback' => 'joinPoster',
```
[the remark this answers is not in the document]
Comment (attiks) [reply html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php:207]: Is the poster the owner of the node?

[the remark this answers is not in the document]
Comment (orakili) [reply html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php:298]: "Me" is the owner of the node.
"Other" is anybody else that is also allowed to post for the same organizations: https://github.com/UN-OCHA/rwint-site/blob/44935f32acb0fcecf1968c7bfe2d6975b226edd4/site/profiles/reliefweb/modules/features/reliefweb_opportunities/includes/user-posts.inc#L372-L427

**Comment by left23** [comment]
I've added some style-related commits.
I also doubled checked and the user needs to have administrator role to view the user posts page

**Review by orakili — APPROVED** [review]
Added several commits notably to fix the query to use the "my posts" user instead of the current user. Also adjusted a bit the styling and added the permissions.

@attiks, @left23 I think it's ready and I'm approving but could you test a bit, notably the different filters to verify nothing breaks?

**Comment by left23** [comment]
@orakili I'm getting a DB error when I select single dates for the Deadline:
```
The website encountered an unexpected error. Please try again later.

Drupal\Core\Database\DatabaseExceptionWrapper: SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'AND )) AND (("node_field_data"."uid" = '2852740') OR ("user_sources"."field_sour' at line 9: SELECT DISTINCT "subquery"."content_entity_id" AS "entity_id", "subquery"."created" AS "sort" FROM (SELECT "cms"."moderation_state" AS "moderation_state", "cms"."content_entity_id" AS "content_entity_id", "node_field_data"."created" AS "created", CASE cms.moderation_state WHEN 'draft' THEN @draft := @draft + 1 WHEN 'pending' THEN @pending := @pending + 1 WHEN 'published' THEN @published := @published + 1 WHEN 'on_hold' THEN @on_hold := @on_hold + 1 WHEN 'refused' THEN @refused := @refused + 1 WHEN 'expired' THEN @expired := @expired + 1 WHEN 'duplicate' THEN @duplicate := @duplicate + 1 END AS "expression", COALESCE(node__field_job_closing_date.field_job_closing_date_value, node__field_registration_deadline.field_registration_deadline_value) AS "deadline" FROM "content_moderation_state_field_data" "cms" INNER JOIN "node_field_data" "node_field_data" ON node_field_data.nid = cms.content_entity_id INNER JOIN "node_field_data" "node_field_data_2" ON node_field_data_2.nid = node_field_data.nid LEFT OUTER JOIN "node__field_job_closing_date" "node__field_job_closing_date" ON node__field_job_closing_date.entity_id = node_field_data.nid LEFT OUTER JOIN "node__field_registration_deadline" "node__field_registration_deadline" ON node__field_registration_deadline.entity_id = node_field_data.nid LEFT OUTER JOIN "node__field_source" "user_sources" ON user_sources.entity_id = node_field_data.nid WHERE ("cms"."content_entity_type_id" = :db_condition_placeholder_0) AND ("node_field_data"."type" IN (:db_condition_placeholder_1, :db_condition_placeholder_2)) AND ("cms"."moderation_state" IN (:db_condition_placeholder_3, :db_condition_placeholder_4, :db_condition_placeholder_5, :db_condition_placeholder_6, :db_condition_placeholder_7, :db_condition_placeholder_8, :db_condition_placeholder_9)) AND ("node_field_data_2"."type" = :db_condition_placeholder_10) AND ((UNIX_TIMESTAMP(COALESCE(node__field_job_closing_date.field_job_closing_date_value, node__field_registration_deadline.field_registration_deadline_value)) >= AND )) AND (("node_field_data"."uid" = :db_condition_placeholder_11) OR ("user_sources"."field_source_target_id" IN (:db_condition_placeholder_12)))) "subquery" ORDER BY "subquery"."created" DESC LIMIT 30 OFFSET 0; Array ( [:db_condition_placeholder_0] => node [:db_condition_placeholder_1] => job [:db_condition_placeholder_2] => training [:db_condition_placeholder_3] => draft [:db_condition_placeholder_4] => pending [:db_condition_placeholder_5] => published [:db_condition_placeholder_6] => on_hold [:db_condition_placeholder_7] => refused [:db_condition_placeholder_8] => expired [:db_condition_placeholder_9] => duplicate [:db_condition_placeholder_10] => training [:db_condition_placeholder_11] => 2852740 [:db_condition_placeholder_12] => 1503 ) in Drupal\reliefweb_moderation\ModerationServiceBase->executeQuery() (line 1157 of modules/custom/reliefweb_moderation/src/ModerationServiceBase.php).
Drupal\Core\Database\StatementWrapper->execute(Array, Array) (Line: 874)
Drupal\Core\Database\Connection->query('SELECT DISTINCT "subquery"."content_entity_id" AS "entity_id", "subquery"."created" AS "sort"
FROM
(SELECT "cms"."moderation_state" AS "moderation_state", "cms"."content_entity_id" AS "content_entity_id", "node_field_data"."created" AS "created", CASE cms.moderation_state WHEN 'draft' THEN @draft := @draft + 1 WHEN 'pending' THEN @pending := @pending + 1 WHEN 'published' THEN @published := @published + 1 WHEN 'on_hold' THEN @on_hold := @on_hold + 1 WHEN 'refused' THEN @refused := @refused + 1 WHEN 'expired' THEN @expired := @expired + 1 WHEN 'duplicate' THEN @duplicate := @duplicate + 1  END AS "expression", COALESCE(node__field_job_closing_date.field_job_closing_date_value, node__field_registration_deadline.field_registration_deadline_value) AS "deadline"
FROM
{content_moderation_state_field_data} "cms"
INNER JOIN {node_field_data} "node_field_data" ON node_field_data.nid = cms.content_entity_id
INNER JOIN {node_field_data} "node_field_data_2" ON node_field_data_2.nid = node_field_data.nid
LEFT OUTER JOIN {node__field_job_closing_date} "node__field_job_closing_date" ON node__field_job_closing_date.entity_id = node_field_data.nid
LEFT OUTER JOIN {node__field_registration_deadline} "node__field_registration_deadline" ON node__field_registration_deadline.entity_id = node_field_data.nid
LEFT OUTER JOIN {node__field_source} "user_sources" ON user_sources.entity_id = node_field_data.nid
WHERE ("cms"."content_entity_type_id" = :db_condition_placeholder_0) AND ("node_field_data"."type" IN (:db_condition_placeholder_1, :db_condition_placeholder_2)) AND ("cms"."moderation_state" IN (:db_condition_placeholder_3, :db_condition_placeholder_4, :db_condition_placeholder_5, :db_condition_placeholder_6, :db_condition_placeholder_7, :db_condition_placeholder_8, :db_condition_placeholder_9)) AND ("node_field_data_2"."type" = :db_condition_placeholder_10) AND ((UNIX_TIMESTAMP(COALESCE(node__field_job_closing_date.field_job_closing_date_value, node__field_registration_deadline.field_registration_deadline_value)) &gt;=  AND )) AND (("node_field_data"."uid" = :db_condition_placeholder_11) OR ("user_sources"."field_source_target_id" IN (:db_condition_placeholder_12)))) "subquery"
ORDER BY "subquery"."created" DESC
LIMIT 30 OFFSET 0', Array, Array) (Line: 512)
Drupal\Core\Database\Query\Select->execute() (Line: 1157)
Drupal\reliefweb_moderation\ModerationServiceBase->executeQuery(Array, 30) (Line: 437)
Drupal\reliefweb_moderation\ModerationServiceBase->getTable(Array, 30) (Line: 126)
Drupal\reliefweb_user_posts\Controller\UserPostsPage->getContent(Object, Object)
call_user_func_array(Array, Array) (Line: 123)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}() (Line: 578)
Drupal\Core\Render\Renderer->executeInRenderContext(Object, Object) (Line: 124)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->wrapControllerExecutionInRenderContext(Array, Array) (Line: 97)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}() (Line: 158)
Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object, 1) (Line: 80)
Symfony\Component\HttpKernel\HttpKernel->handle(Object, 1, 1) (Line: 57)
Drupal\Core\StackMiddleware\Session->handle(Object, 1, 1) (Line: 47)
Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object, 1, 1) (Line: 106)
Drupal\page_cache\StackMiddleware\PageCache->pass(Object, 1, 1) (Line: 85)
Drupal\page_cache\StackMiddleware\PageCache->handle(Object, 1, 1) (Line: 47)
Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object, 1, 1) (Line: 52)
Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object, 1, 1) (Line: 23)
Stack\StackedHttpKernel->handle(Object, 1, 1) (Line: 717)
Drupal\Core\DrupalKernel->handle(Object) (Line: 19)

And when I select a date range I get a warning:
```
Warning: Trying to access array offset on value of type int in Drupal\reliefweb_user_posts\Services\UserPostsService->conditionDeadline() (line 422 of modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php).
Drupal\reliefweb_user_posts\Services\UserPostsService->conditionDeadline(Array, Object, 'COALESCE(node__field_job_closing_date.field_job_closing_date_value, node__field_registration_deadline.field_registration_deadline_value)', 1636416000, '&gt;=') (Line: 1559)
Drupal\reliefweb_moderation\ModerationServiceBase->addFilterCondition(Array, Object, 'COALESCE(node__field_job_closing_date.field_job_closing_date_value, node__field_registration_deadline.field_registration_deadline_value)', 1636416000, '&gt;=') (Line: 1466)
Drupal\reliefweb_moderation\ModerationServiceBase->filterQuery(Object, Array) (Line: 1138)
Drupal\reliefweb_moderation\ModerationServiceBase->executeQuery(Array, 30) (Line: 437)
Drupal\reliefweb_moderation\ModerationServiceBase->getTable(Array, 30) (Line: 126)
Drupal\reliefweb_user_posts\Controller\UserPostsPage->getContent(Object, Object)
call_user_func_array(Array, Array) (Line: 123)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}() (Line: 578)
Drupal\Core\Render\Renderer->executeInRenderContext(Object, Object) (Line: 124)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->wrapControllerExecutionInRenderContext(Array, Array) (Line: 97)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}() (Line: 158)
Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object, 1) (Line: 80)
Symfony\Component\HttpKernel\HttpKernel->handle(Object, 1, 1) (Line: 57)
Drupal\Core\StackMiddleware\Session->handle(Object, 1, 1) (Line: 47)
Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object, 1, 1) (Line: 106)
Drupal\page_cache\StackMiddleware\PageCache->pass(Object, 1, 1) (Line: 85)
Drupal\page_cache\StackMiddleware\PageCache->handle(Object, 1, 1) (Line: 47)
Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object, 1, 1) (Line: 52)
Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object, 1, 1) (Line: 23)
Stack\StackedHttpKernel->handle(Object, 1, 1) (Line: 717)
Drupal\Core\DrupalKernel->handle(Object) (Line: 19)
```


I'm not sure how the Author filter is suppose to work. I'm testing locally with content created by `lisacorcoran` user and when I enter that into the Author input, I'm not able to press enter as if the autocomplete isn't connecting. Same result with the admin user (I'm assuming Author = `username` - In these cases, both `username` and `display name` are the same)

The other Omnibox filters seem to work as I would expect.

When I am at  `/user/2852740/posts` (logged in as that user) and I have none of the Type or Posted by checkboxes selected, I get all the content, but if I check both "Me" and "Other" for Posted by, I get no results. I find this a little counter intuitive. Same with having both "Job" and "Training" selected. 
On D7 it seems to work slightly differently but I'm having trouble putting my finger on it because there's a lack of content examples and I don't know the expectations. For example, should it matter that I am on another user's post page versus my own?

**Comment by orakili** [comment]
@left23 I think your local version is outdated. There is no author filter anymore for example. The filters should look like that:
Screen Shot 2021-11-26 at 9 02 52

The latest commit is 511c0fbdfe4594ebb45dabb1a5e6249a2856b8a3

Could you confirm?

**Comment by left23** [comment]
@orakili Yes, sorry, I didn't update my local branch. I double checked everything in my last comment and all the issues I point out are irrelevant.

**Review by left23 — APPROVED** [review]

**Pull request merged**